sql/moz-fx-data-marketing-prod/ga_derived/downloads_with_attribution_v2/query.sql

```diff
@@ -0,0 +1,404 @@
+    -- download_session_id per GA session implementation will be left as is.
+    mozfun.stats.mode_last(ARRAY_AGG(download_session_id)) AS download_session_id,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(landing_page)) AS landing_page,
+    LOGICAL_OR(has_ga_download_event) AS has_ga_download_event
+  FROM
+    all_hits
+  WHERE
+     hit_type = 'EVENT'
+  GROUP BY
+    client_id,
+    visit_id
+    -- download_session_id  cannot group by download_session_id here since not every hit will have download_session_id
+),
+ga_session_dimensions AS (
+  SELECT
+    clientId AS client_id,
+    visitId AS visit_id,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(geoNetwork.country)) AS country,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(trafficSource.adContent)) AS ad_content,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(trafficSource.campaign)) AS campaign,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(trafficSource.medium)) AS medium,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(trafficSource.source)) AS source,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(device.deviceCategory)) AS device_category,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(device.operatingSystem)) AS os,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(device.browser)) AS browser,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(device.browserVersion)) AS browser_version,
+    mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(device.language)) AS `language`,
+    IFNULL(mozfun.stats.mode_last_retain_nulls(ARRAY_AGG(totals.timeOnSite)), 0) AS time_on_site
+  FROM
+    `moz-fx-data-marketing-prod.65789850.ga_sessions_*` AS ga
+  WHERE
+    _TABLE_SUFFIX
+    BETWEEN FORMAT_DATE('%Y%m%d', DATE_SUB(@download_date, INTERVAL 2 DAY))
+    AND FORMAT_DATE('%Y%m%d', DATE_ADD(@download_date, INTERVAL 1 DAY))
+  GROUP BY
+    client_id,
+    visit_id
+),
+ga_sessions_with_hits_fields AS (
+  SELECT
+    * EXCEPT (pageviews, unique_pageviews),
+    IFNULL(pageviews, 0) AS pageviews,
+    IFNULL(unique_pageviews, 0) AS unique_pageviews,
+  FROM
+    ga_session_dimensions ga
+  LEFT JOIN
+    page_hits p
+  USING
+    (client_id, visit_id)
+  JOIN
+    event_hits
+  USING
+    (client_id, visit_id)
+),
+
+-- Extract all the download rows, de-duping and tracking number of duplicates per download token.
+stub_downloads AS (
+  SELECT
+    s.jsonPayload.fields.visit_id AS stub_visit_id,
+    jsonPayload.fields.session_id AS stub_download_session_id,
+    s.jsonPayload.fields.dltoken AS dltoken,
+    (COUNT(*) - 1) AS count_dltoken_duplicates,
+    DATE(@download_date) AS download_date
+  FROM
+    `moz-fx-stubattribut-prod-32a5.stubattribution_prod.stdout` s
+  WHERE
+    DATE(s.timestamp) = @download_date
+    AND s.jsonPayload.fields.log_type = 'download_started'
+  GROUP BY
+    stub_visit_id,
+    stub_download_session_id,
+    dltoken
+),
+
+multiple_downloads_in_session AS (
+  SELECT
+    stub_visit_id,
+    stub_download_session_id,
+    CASE
```

Comment:
Could this be expressed as `IF()`?

Reply:
Yep  :-)  I've been in the weeds on this query for so long that I don't see these things.